cluster/src/shard_operator.rs

```diff
@@ -0,0 +1,404 @@
+
+        let table_defs = tables
+            .into_iter()
+            .map(|info| TableDef {
+                catalog_name: ctx.catalog.clone(),
+                schema_name: info.schema_name,
+                id: TableId::from(info.id),
+                name: info.name,
+            })
+            .collect();
+
+        let open_shard_request = catalog::schema::OpenShardRequest {
+            shard_id: shard_info.id,
+            table_defs,
+            engine: ctx.engine,
+        };
+        let opts = OpenOptions {
+            table_engine: ctx.table_engine.clone(),
+        };
+
+        ctx.table_operator
+            .open_shard(open_shard_request, opts)
+            .await
+            .box_err()
+            .with_context(|| OpenShardWithCause {
+                msg: format!("shard_info:{shard_info:?}"),
+            })?;
+
+        info!("ShardOperator open sequentially finish, shard_id:{shard_info:?}");
+
+        Ok(())
+    }
+
+    pub async fn close(&self, ctx: CloseContext) -> Result<()> {
+        let (shard_info, tables) = {
+            let data = self.data.read().unwrap();
+            let shard_info = data.shard_info.clone();
+            let tables = data.tables.clone();
+
+            (shard_info, tables)
+        };
+        info!("ShardOperator close sequentially begin, shard_info:{shard_info:?}");
+
+        {
+            let mut data = self.data.write().unwrap();
+            data.freeze();
+            info!("Shard is frozen before closed, shard_id:{}", shard_info.id);
+        }
+
+        let table_defs = tables
+            .into_iter()
+            .map(|info| TableDef {
+                catalog_name: ctx.catalog.clone(),
+                schema_name: info.schema_name,
+                id: TableId::from(info.id),
+                name: info.name,
+            })
+            .collect();
+        let close_shard_request = catalog::schema::CloseShardRequest {
+            shard_id: shard_info.id,
+            table_defs,
+            engine: ctx.engine,
+        };
+        let opts = CloseOptions {
+            table_engine: ctx.table_engine,
+        };
+
+        ctx.table_operator
+            .close_shard(close_shard_request, opts)
+            .await
+            .box_err()
+            .with_context(|| CloseShardWithCause {
+                msg: format!("shard_info:{shard_info:?}"),
+            })?;
+
+        // Try to close wal region
+        ctx.wal_region_closer
+            .close_region(shard_info.id)
+            .await
+            .with_context(|| CloseShardWithCause {
+                msg: format!("shard_info:{shard_info:?}"),
+            })?;
+
+        info!("ShardOperator close sequentially finish, shard_info:{shard_info:?}");
+
+        Ok(())
+    }
+
+    pub async fn create_table(&self, ctx: CreateTableContext) -> Result<()> {
+        let shard_info = ctx.updated_table_info.shard_info.clone();
+        let table_info = ctx.updated_table_info.table_info.clone();
```

Comment:
Are these two clones necessary?

Reply:
Maybe necessary now? Becasue we need to use these two inner struct, after moving `ctx.updated_table_info`.